lib/proxy/peer/quic/quic_test.go

```diff
@@ -0,0 +1,414 @@
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			serverTransport := newTransport(t)
+			server, err := NewServer(ServerConfig{
+				Dialer: noDialer,
+				GetCertificate: func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+					return &tc.serverCert, nil
+				},
+				GetClientCAs: func(*tls.ClientHelloInfo) (*x509.CertPool, error) {
+					return tc.clientCAs, nil
+				},
+			})
+			require.NoError(t, err)
+			go server.Serve(serverTransport)
+			t.Cleanup(func() { _ = server.Close() })
+
+			client, err := NewClientConn(ClientConnConfig{
+				Log:         slog.Default(),
+				PeerID:      "correctproxy",
+				PeerAddr:    serverTransport.Conn.LocalAddr().String(),
+				ClusterName: "testcluster",
+				GetTLSCertificate: func() (*tls.Certificate, error) {
+					return &tc.clientCert, nil
+				},
+				GetTLSRoots: func() (*x509.CertPool, error) {
+					return tc.rootCAs, nil
+				},
+				Transport: clientTransport,
+			})
+			require.NoError(t, err)
+			t.Cleanup(func() { _ = client.Close() })
+
+			err = client.Ping(context.Background())
+			tc.check(t, err)
+		})
+	}
+}
+
+func TestBasic(t *testing.T) {
+	t.Parallel()
+
+	hostCA := newCA(t)
+	hostCAPool := x509.NewCertPool()
+	hostCAPool.AddCert(hostCA.Cert)
+
+	serverHostID := uuid.NewString()
+	serverCert := newCert(t, hostCA, tlsca.Identity{
+		Username: serverHostID + ".testcluster",
+		Groups:   []string{string(types.RoleProxy)},
+	})
+	clientHostID := uuid.NewString()
+	clientCert := newCert(t, hostCA, tlsca.Identity{
+		Username: clientHostID + ".testcluster",
+		Groups:   []string{string(types.RoleProxy)},
+	})
+
+	serverTransport := newTransport(t)
+	server, err := NewServer(ServerConfig{
+		Dialer: noDialer,
+		GetCertificate: func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+			return &serverCert, nil
+		},
+		GetClientCAs: func(*tls.ClientHelloInfo) (*x509.CertPool, error) {
+			return hostCAPool, nil
+		},
+	})
+	require.NoError(t, err)
+	go server.Serve(serverTransport)
+	t.Cleanup(func() { _ = server.Close() })
+
+	client, err := NewClientConn(ClientConnConfig{
+		Log:         slog.Default(),
+		PeerID:      serverHostID,
+		PeerAddr:    serverTransport.Conn.LocalAddr().String(),
+		ClusterName: "testcluster",
+		GetTLSCertificate: func() (*tls.Certificate, error) {
+			return &clientCert, nil
+		},
+		GetTLSRoots: func() (*x509.CertPool, error) {
+			return hostCAPool, nil
+		},
+		Transport: newTransport(t),
+	})
+	require.NoError(t, err)
+	t.Cleanup(func() { _ = client.Close() })
+
+	require.NoError(t, client.Ping(context.Background()))
+
+	conn, err := client.Dial(
+		uuid.NewString()+".testcluster",
+		&utils.NetAddr{
+			AddrNetwork: "tcp",
+			Addr:        "1.2.3.4:56",
+		}, &utils.NetAddr{
+			AddrNetwork: "tcp",
+			Addr:        "7.8.9.0:12",
+		},
+		types.NodeTunnel,
+	)
+	if err == nil {
+		_ = conn.Close()
```

Comment:
It's not supposed to happen, but it doesn't feel great to leak an `io.Closer` just because the testing logic has decided to terminate the test early. Replaced the check with an `assert.Error` instead.

Reply:
Yep, a t.Error or something like that helps a lot in debugging if/when the impossible happens.